File: app/src/main/java/com/voltpay/app/ui/contacts/ContactsActivity.java
package com.voltpay.app.ui.contacts;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.PopupMenu;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.work.Constraints;
import androidx.work.NetworkType;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.voltpay.app.R;
import com.voltpay.app.data.db.ContactDao;
import com.voltpay.app.data.db.DatabaseHelper;
import com.voltpay.app.data.db.SyncItemDao;
import com.voltpay.app.data.model.Contact;
import com.voltpay.app.data.model.SyncItem;
import com.voltpay.app.sync.SyncWorker;
import com.voltpay.app.ui.payment.SendMoneyActivity;

import java.util.List;

public class ContactsActivity extends AppCompatActivity {

    private ContactDao contactDao;
    private SyncItemDao syncItemDao;
    private ContactsAdapter adapter;
    private View emptyStateLayout;
    private RecyclerView rvContacts;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_contacts);

        contactDao = new ContactDao(this);
        syncItemDao = new SyncItemDao(this);

        emptyStateLayout = findViewById(R.id.emptyStateLayout);
        rvContacts = findViewById(R.id.rvContacts);
        EditText etSearch = findViewById(R.id.etSearch);
        ImageView btnBack = findViewById(R.id.btnBack);
        View fabAddContact = findViewById(R.id.fabAddContact);
        View btnAddContactEmpty = findViewById(R.id.btnAddContactEmpty);

        btnBack.setOnClickListener(v -> finish());
        
        View.OnClickListener addContactListener = v -> {
            Intent intent = new Intent(this, AddEditContactActivity.class);
            startActivity(intent);
        };
        
        if (fabAddContact != null) fabAddContact.setOnClickListener(addContactListener);
        if (btnAddContactEmpty != null) btnAddContactEmpty.setOnClickListener(addContactListener);

        setupRecyclerView();

        etSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                searchContacts(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        loadContacts();
    }

    private void setupRecyclerView() {
        rvContacts.setLayoutManager(new LinearLayoutManager(this));
        adapter = new ContactsAdapter(new ContactsAdapter.OnContactClickListener() {
            @Override
            public void onContactClick(Contact contact) {
                Intent intent = new Intent(ContactsActivity.this, SendMoneyActivity.class);
                intent.putExtra("EXTRA_UPI_ID", contact.getUpiId());
                startActivity(intent);
            }

            @Override
            public void onContactLongClick(Contact contact, View view) {
                showContextMenu(contact, view);
            }
        });
        rvContacts.setAdapter(adapter);
    }

    private void loadContacts() {
        List<Contact> contacts = contactDao.getAllContacts();
        updateUI(contacts);
    }

    private void searchContacts(String query) {
        if (query.trim().isEmpty()) {
            loadContacts();
        } else {
            List<Contact> contacts = contactDao.searchByName(query.trim());
            updateUI(contacts);
        }
    }

    private void updateUI(List<Contact> contacts) {
        if (contacts.isEmpty()) {
            rvContacts.setVisibility(View.GONE);
            emptyStateLayout.setVisibility(View.VISIBLE);
        } else {
            rvContacts.setVisibility(View.VISIBLE);
            emptyStateLayout.setVisibility(View.GONE);
            adapter.setContacts(contacts);
        }
    }

    private void showContextMenu(Contact contact, View view) {
        PopupMenu popupMenu = new PopupMenu(this, view);
        popupMenu.getMenu().add(0, 1, 0, "Edit");
        popupMenu.getMenu().add(0, 2, 1, "Delete");

        popupMenu.setOnMenuItemClickListener(item -> {
            if (item.getItemId() == 1) {
                Intent intent = new Intent(this, AddEditContactActivity.class);
                intent.putExtra("EXTRA_CONTACT_ID", contact.getId());
                startActivity(intent);
                return true;
            } else if (item.getItemId() == 2) {
                showDeleteConfirmation(contact);
                return true;
            }
            return false;
        });
        popupMenu.show();
    }

    private void showDeleteConfirmation(Contact contact) {
        new AlertDialog.Builder(this)
                .setTitle("Delete Contact")
                .setMessage("Delete " + contact.getName() + "? This cannot be undone.")
                .setPositiveButton("Delete", (dialog, which) -> {
                    contactDao.deleteContact(contact.getId());
                    queueSync();
                    loadContacts();
                })
                .setNegativeButton("Cancel", null)
                .show();
    }

    private void queueSync() {
        long currentTime = System.currentTimeMillis();
        SyncItem existingItem = syncItemDao.getPendingItemByActionType("SYNC_CONTACTS");

        if (existingItem != null) {
            syncItemDao.updateSyncItemPayload(existingItem.getId(), "update_triggered", currentTime);
        } else {
            SyncItem newItem = new SyncItem();
            newItem.setActionType("SYNC_CONTACTS");
            newItem.setPayload("update_triggered");
            newItem.setCreatedAt(currentTime);
            newItem.setStatus("PENDING");
            syncItemDao.insertSyncItem(newItem);
        }

        Constraints constraints = new Constraints.Builder()
                .setRequiredNetworkType(NetworkType.CONNECTED)
                .build();
        
        OneTimeWorkRequest syncRequest = new OneTimeWorkRequest.Builder(SyncWorker.class)
                .setConstraints(constraints)
                .build();
                
        WorkManager.getInstance(this).enqueue(syncRequest);
    }
}

File: app/src/main/java/com/voltpay/app/ui/contacts/ContactsActivity.kt
package com.voltpay.app.ui.contacts

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.PopupMenu
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.work.Constraints
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequest
import androidx.work.WorkManager
import com.voltpay.app.R
import com.voltpay.app.data.db.ContactDao
import com.voltpay.app.data.db.SyncItemDao
import com.voltpay.app.data.model.Contact
import com.voltpay.app.data.model.SyncItem
import com.voltpay.app.sync.SyncWorker
import com.voltpay.app.ui.payment.SendMoneyActivity

class ContactsActivity : AppCompatActivity() {

    private lateinit var contactDao: ContactDao
    private lateinit var syncItemDao: SyncItemDao
    private lateinit var adapter: ContactsAdapter
    private lateinit var emptyStateLayout: View
    private lateinit var rvContacts: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_contacts)

        contactDao = ContactDao(this)
        syncItemDao = SyncItemDao(this)

        emptyStateLayout = findViewById(R.id.emptyStateLayout)
        rvContacts = findViewById(R.id.rvContacts)
        val etSearch = findViewById<EditText>(R.id.etSearch)
        
        findViewById<ImageView>(R.id.btnBack).setOnClickListener { finish() }
        
        val addContactListener = View.OnClickListener {
            val intent = Intent(this, AddEditContactActivity::class.java)
            startActivity(intent)
        }
        
        findViewById<View>(R.id.fabAddContact)?.setOnClickListener(addContactListener)
        findViewById<View>(R.id.btnAddContactEmpty)?.setOnClickListener(addContactListener)

        setupRecyclerView()

        etSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                searchContacts(s.toString())
            }
            override fun afterTextChanged(s: Editable) {}
        })
    }

    override fun onResume() {
        super.onResume()
        loadContacts()
    }

    private fun setupRecyclerView() {
        rvContacts.layoutManager = LinearLayoutManager(this)
        adapter = ContactsAdapter(object : ContactsAdapter.OnContactClickListener {
            override fun onContactClick(contact: Contact) {
                val intent = Intent(this@ContactsActivity, SendMoneyActivity::class.java)
                intent.putExtra("EXTRA_UPI_ID", contact.upiId)
                startActivity(intent)
            }

            override fun onContactLongClick(contact: Contact, view: View) {
                showContextMenu(contact, view)
            }
        })
        rvContacts.adapter = adapter
    }

    private fun loadContacts() {
        val contacts = contactDao.getAllContacts()
        updateUI(contacts)
    }

    private fun searchContacts(query: String) {
        if (query.trim().isEmpty()) {
            loadContacts()
        } else {
            val contacts = contactDao.searchByName(query.trim())
            updateUI(contacts)
        }
    }

    private fun updateUI(contacts: List<Contact>) {
        if (contacts.isEmpty()) {
            rvContacts.visibility = View.GONE
            emptyStateLayout.visibility = View.VISIBLE
        } else {
            rvContacts.visibility = View.VISIBLE
            emptyStateLayout.visibility = View.GONE
            adapter.setContacts(contacts)
        }
    }

    private fun showContextMenu(contact: Contact, view: View) {
        val popupMenu = PopupMenu(this, view)
        popupMenu.menu.add(0, 1, 0, "Edit")
        popupMenu.menu.add(0, 2, 1, "Delete")

        popupMenu.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                1 -> {
                    val intent = Intent(this, AddEditContactActivity::class.java)
                    intent.putExtra("EXTRA_CONTACT_ID", contact.id)
                    startActivity(intent)
                    true
                }
                2 -> {
                    showDeleteConfirmation(contact)
                    true
                }
                else -> false
            }
        }
        popupMenu.show()
    }

    private fun showDeleteConfirmation(contact: Contact) {
        AlertDialog.Builder(this)
            .setTitle("Delete Contact")
            .setMessage("Delete ${contact.name}? This cannot be undone.")
            .setPositiveButton("Delete") { _, _ ->
                contactDao.deleteContact(contact.id)
                queueSync()
                loadContacts()
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun queueSync() {
        val currentTime = System.currentTimeMillis()
        val existingItem = syncItemDao.getPendingItemByActionType("SYNC_CONTACTS")

        if (existingItem != null) {
            syncItemDao.updateSyncItemPayload(existingItem.id, "update_triggered", currentTime)
        } else {
            val newItem = SyncItem()
            newItem.actionType = "SYNC_CONTACTS"
            newItem.payload = "update_triggered"
            newItem.createdAt = currentTime
            newItem.status = "PENDING"
            syncItemDao.insertSyncItem(newItem)
        }

        val constraints = Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build()
        
        val syncRequest = OneTimeWorkRequest.Builder(SyncWorker::class.java)
            .setConstraints(constraints)
            .build()
            
        WorkManager.getInstance(this).enqueue(syncRequest)
    }
}
